### PR TITLE
Use spack-stack and switch to tags/commit hashes for JEDI repos

### LINKED ISSUE #142: Switch to using spack-stack for Hera and Orion
Once spack-stack is installed on Hera, change the modulefiles to use spack-stack on both HPC platforms. Note: this may require some double-checking that the existing python virtual environments are still working as intended.

### PR DESCRIPTION
**Pull request by CoryMartin-NOAA**
This PR closes issue #143 and closes #142 

This PR does 3 main things:
1 - Switches the modulefiles to use the 'spack-stack' for both Hera and Orion
2 - Wraps the module load part of the build script in `set +e` and `set -e` so that it will not error on `$PS1 Unbound Variable`
3 - Adds commit hashes to most of the JEDI repos so that we can control which version is being checked out (need to discuss more with @RussTreadon-NOAA and @guillaumevernieres about how to maintain this moving forward, but this is needed ASAP to fix issues with global-workflow)

### CODE REVIEW

**Comment by CoryMartin-NOAA** [comment]
`./build.sh -t orion` and `./build.sh -t hera` should work on their respective platforms, but please try it out to ensure my environment is not special.

**Comment by CoryMartin-NOAA** [comment]
@LiaofanLin-NOAA this might help you be able to build now on Hera

**Comment by CoryMartin-NOAA** [comment]
Marking as draft because I thought it was working properly on both platforms but that seems to not be the case... may need to end up using their `spack-stack` python for GDASApp...

**Review by guillaumevernieres — DISMISSED** [review]
I'm fine with tags in develop for now until we come up with a better solution.
Preemptively approving without waiting for the build to complete on orion ...

**Comment by RussTreadon-NOAA** [comment]
A fresh clone of GDASApp branch `feature/spack-stack-tags` has been installed on Orion.  `./build.sh -t orion -v` ran to completion.   `ctest -N test_gdasapp` ran 14 tests.   8 tests failed.

```
The following tests passed:
        test_gdasapp_check_python_norms
        test_gdasapp_genYAML_prep
        test_gdasapp_convert_ewok_yaml
        test_gdasapp_soca_obsdb
        test_gdasapp_convert_gsi_satbias
        test_gdasapp_store_gsi_satbias

43% tests passed, 8 tests failed out of 14

Total Test time (real) =  61.07 sec

The following tests FAILED:
        1590 - test_gdasapp_check_yaml_keys (Failed)
        1591 - test_gdasapp_check_valid_yaml (Failed)
        1592 - test_gdasapp_jedi_increment_to_fv3 (Failed)
        1593 - test_gdasapp_generate_yaml_func (Failed)
        1595 - test_gdasapp_genYAML_run (Failed)
        1598 - test_gdasapp_soca_ana_prep (Failed)
        1599 - test_gdasapp_soca_ana_run (Not Run)
        1600 - test_gdasapp_convert_bufr_adpsfc (Subprocess aborted)
```
 
This is surprising.   Previous execution of test_gdsaapp resulted in only two failures
```
The following tests FAILED:
        1557 - test_gdasapp_soca_ana_prep (Failed)
        1558 - test_gdasapp_soca_ana_run (Not Run)
```

**Comment by CoryMartin-NOAA** [comment]
Thanks @RussTreadon-NOAA yeah I think there is something wrong with the python installation. The code builds but I need to investigate this issue further.

**Comment by CoryMartin-NOAA** [comment]
I'm wondering if `test_gdasapp_convert_bufr_adpsfc` is related to changes in ioda-converters and not a result of this PR? Regardless, if so, we need to update the test (probably YAML) to make it pass.

**Comment by CoryMartin-NOAA** [comment]
@RussTreadon-NOAA do you want #147 in a separate PR or part of this one? The test failure of `test_gdasapp_convert_bufr_adpsfc` can be solved by using a modified YAML file to match what now in ioda-converters.

**Comment by RussTreadon-NOAA** [comment]
@CoryMartin-NOAA , since #147 is related to this PR I'd say let's bundle them together.

**Comment by guillaumevernieres** [comment]
@CoryMartin-NOAA and @RussTreadon-NOAA , you can ignore the soca tests, maybe  the prep test should work, but I never updated the yamls for `test_gdasapp_soca_ana_run`, so that one should fail. On my todo list ...

**Comment by CoryMartin-NOAA** [comment]
```
Test project /work2/noaa/stmp/cmartin/tmp2/gdasapp3/build
      Start 1598: test_gdasapp_check_python_norms
 1/14 Test #1598: test_gdasapp_check_python_norms ......   Passed    1.90 sec
      Start 1599: test_gdasapp_check_yaml_keys
 2/14 Test #1599: test_gdasapp_check_yaml_keys .........   Passed    1.07 sec
      Start 1600: test_gdasapp_check_valid_yaml
 3/14 Test #1600: test_gdasapp_check_valid_yaml ........   Passed    0.70 sec
      Start 1601: test_gdasapp_jedi_increment_to_fv3
 4/14 Test #1601: test_gdasapp_jedi_increment_to_fv3 ...   Passed   13.89 sec
      Start 1602: test_gdasapp_generate_yaml_func
 5/14 Test #1602: test_gdasapp_generate_yaml_func ......   Passed    7.24 sec
      Start 1603: test_gdasapp_genYAML_prep
 6/14 Test #1603: test_gdasapp_genYAML_prep ............   Passed    0.03 sec
      Start 1604: test_gdasapp_genYAML_run
 7/14 Test #1604: test_gdasapp_genYAML_run .............   Passed    1.15 sec
      Start 1605: test_gdasapp_convert_ewok_yaml
 8/14 Test #1605: test_gdasapp_convert_ewok_yaml .......   Passed    0.17 sec
      Start 1606: test_gdasapp_soca_obsdb
 9/14 Test #1606: test_gdasapp_soca_obsdb ..............   Passed    0.67 sec
      Start 1607: test_gdasapp_soca_ana_prep
10/14 Test #1607: test_gdasapp_soca_ana_prep ...........***Failed    0.09 sec
      Start 1608: test_gdasapp_soca_ana_run
Failed to change working directory to /work2/noaa/stmp/cmartin/tmp2/gdasapp3/build/test/soca/3dvar/analysis : No such file or directory
11/14 Test #1608: test_gdasapp_soca_ana_run ............***Not Run   0.00 sec
      Start 1609: test_gdasapp_convert_bufr_adpsfc
12/14 Test #1609: test_gdasapp_convert_bufr_adpsfc .....   Passed   13.95 sec
      Start 1610: test_gdasapp_convert_gsi_satbias
13/14 Test #1610: test_gdasapp_convert_gsi_satbias .....   Passed    4.18 sec
      Start 1611: test_gdasapp_store_gsi_satbias
14/14 Test #1611: test_gdasapp_store_gsi_satbias .......   Passed    0.57 sec

86% tests passed, 2 tests failed out of 14

Total Test time (real) =  46.25 sec

The following tests FAILED:
	1607 - test_gdasapp_soca_ana_prep (Failed)
	1608 - test_gdasapp_soca_ana_run (Not Run)
```

I think I've fixed the python issue on Orion

**Comment by ADCollard** [comment]
I was able to reproduce the ctest run that Cory made (so only the SOCA tests failing for GDASApp ctests):

```(gdasapp) Orion-login-1[54] acollard$ ctest -R test_gdasapp
Test project /work/noaa/da/acollard/git/GDASApp_spac/build
      Start 1589: test_gdasapp_check_python_norms
 1/14 Test #1589: test_gdasapp_check_python_norms ......   Passed    0.50 sec
      Start 1590: test_gdasapp_check_yaml_keys
 2/14 Test #1590: test_gdasapp_check_yaml_keys .........   Passed    0.09 sec
      Start 1591: test_gdasapp_check_valid_yaml
 3/14 Test #1591: test_gdasapp_check_valid_yaml ........   Passed    0.55 sec
      Start 1592: test_gdasapp_jedi_increment_to_fv3
 4/14 Test #1592: test_gdasapp_jedi_increment_to_fv3 ...   Passed    0.51 sec
      Start 1593: test_gdasapp_generate_yaml_func
 5/14 Test #1593: test_gdasapp_generate_yaml_func ......   Passed    1.10 sec
      Start 1594: test_gdasapp_genYAML_prep
 6/14 Test #1594: test_gdasapp_genYAML_prep ............   Passed    0.02 sec
      Start 1595: test_gdasapp_genYAML_run
 7/14 Test #1595: test_gdasapp_genYAML_run .............   Passed    1.16 sec
      Start 1596: test_gdasapp_convert_ewok_yaml
 8/14 Test #1596: test_gdasapp_convert_ewok_yaml .......   Passed    0.18 sec
      Start 1597: test_gdasapp_soca_obsdb
 9/14 Test #1597: test_gdasapp_soca_obsdb ..............   Passed    0.82 sec
      Start 1598: test_gdasapp_soca_ana_prep
10/14 Test #1598: test_gdasapp_soca_ana_prep ...........***Failed    0.11 sec
      Start 1599: test_gdasapp_soca_ana_run
Failed to change working directory to /work/noaa/da/acollard/git/GDASApp_spac/build/test/soca/3dvar/analysis : No such file or directory
11/14 Test #1599: test_gdasapp_soca_ana_run ............***Not Run   0.00 sec
      Start 1600: test_gdasapp_convert_bufr_adpsfc
12/14 Test #1600: test_gdasapp_convert_bufr_adpsfc .....   Passed    5.84 sec
      Start 1601: test_gdasapp_convert_gsi_satbias
13/14 Test #1601: test_gdasapp_convert_gsi_satbias .....   Passed    5.37 sec
      Start 1602: test_gdasapp_store_gsi_satbias
14/14 Test #1602: test_gdasapp_store_gsi_satbias .......   Passed    0.99 sec

86% tests passed, 2 tests failed out of 14

Total Test time (real) =  17.68 sec

The following tests FAILED:
	1598 - test_gdasapp_soca_ana_prep (Failed)
	1599 - test_gdasapp_soca_ana_run (Not Run)
Errors while running CTest
Output from these tests are in: /work/noaa/da/acollard/git/GDASApp_spac/build/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
```

I will approve this pull request but will continue to test.

**Review by ADCollard — DISMISSED** [review]

**Comment by ADCollard** [comment]
I may have spoken too soon.  Running `/work/noaa/da/acollard/git/GDASApp_spac/ush/ufoeval/run_ufo_hofx_test.sh` (only changing the GDASApp path to point at the current build, not Cory's) I get the following error:

```
Run: Starting oops::Test running test::ObsFilters
Running 1 tests:
Running case 0: ufo/ObsFilters/testFilters ...
Test "ufo/ObsFilters/testFilters" failed with unhandled eckit::Exception: Bad parameter: /observations/./obs space/obsdatain: Mandatory parameter 'engine' not found @  (/work/noaa/da/acollard/git/GDASApp_spac/oops/src/oops/util/parameters/RequiredParameter.h +135 deserialize)
```

**Comment by CoryMartin-NOAA** [comment]
@ADCollard that is because your/my YAML is old/outdated for the current JEDI version/build. It is in the `obsdatain` area, now we need to specify the 'engine'

File: modulefiles/GDAS/hera.lua
Comment (guillaumevernieres) [inline modulefiles/GDAS/hera.lua:43]: @CoryMartin-NOAA , nco is missing, that one of the reasons for the soca test failures.

**Review by guillaumevernieres — DISMISSED** [review]
@CoryMartin-NOAA , @RussTreadon-NOAA Plenty of issues on the soca side ... You can ignore for now, the fix is more involved than just a few lines.

**Comment by ADCollard** [comment]
> @ADCollard that is because your/my YAML is old/outdated for the current JEDI version/build. It is in the `obsdatain` area, now we need to specify the 'engine'

@CoryMartin-NOAA Ah I remember now!  So we need to update the sample files in `parm/atm/obs/testing` ?  I pushed a change to `amsua_n19.yaml` which seems to work.  So `obsdatain` and `obsdataout` need to change but not `geovals` or the bias correction files?

**Comment by CoryMartin-NOAA** [comment]
@ADCollard correct, thanks!

**Comment by CoryMartin-NOAA** [comment]
@guillaumevernieres `nco` has been added back to the modulefiles. I think @RussTreadon-NOAA also wants to add `splib` as well. Once those are added, and we know this builds, I think it is ready for final review/merge.

**Review by ADCollard — DISMISSED** [review]

**Comment by CoryMartin-NOAA** [comment]
Thanks to @guillaumevernieres 's *bon* work, all `gdasapp` tests pass for me on Hera, but on Orion one fails . Looking more closely, I feel like the `run` test fails because the `prep` test, while passing, doesn't work as intended. @guillaumevernieres can you take a look at this with a fresh clone/build?

**Comment by guillaumevernieres** [comment]
> Thanks to @guillaumevernieres 's _bon_ work, all `gdasapp` tests pass for me on Hera, but on Orion one fails . Looking more closely, I feel like the `run` test fails because the `prep` test, while passing, doesn't work as intended. @guillaumevernieres can you take a look at this with a fresh clone/build?

Will look at it right now @CoryMartin-NOAA ... But don't wait for me if you need to merge soon.

**Comment by CoryMartin-NOAA** [comment]
All tests now pass, I think this is ready to go thanks to @guillaumevernieres commenting out the offending tests for now. Just need 1+ approvals

**Review by guillaumevernieres — APPROVED** [review]
Amazing PR @CoryMartin-NOAA!!!

**Review by RussTreadon-NOAA — APPROVED** [review]
Looks good.  Thanks, @CoryMartin-NOAA

**Pull request merged**
